src/plugins/grpc/client.js

```diff
@@ -0,0 +1,143 @@
+'use strict'
+
+const log = require('../../log')
+const pick = require('lodash.pick')
+const Tags = require('../../../ext/tags')
+const TEXT_MAP = require('../../../ext/formats').TEXT_MAP
+
+function getFilter (config, filter) {
+  if (typeof config[filter] === 'function') {
+    return config[filter]
+  }
+
+  if (config[filter] instanceof Array) {
+    return element => pick(element, config[filter])
+  }
+
+  if (config.hasOwnProperty(filter)) {
+    log.error(`Expected '${filter}' to be an array or function.`)
+  }
+
+  return null
+}
+
+function getMethodType (definition) {
+  if (definition.requestStream) {
+    if (definition.responseStream) {
+      return 'bidi'
+    }
+
+    return 'client_stream'
+  }
+
+  if (definition.responseStream) {
+    return 'server_stream'
+  }
+
+  return 'unary'
+}
+
+function patch (grpc, tracer, config) {
+  if (config.client === false) return
+
+  config = config.client || config
+
+  const configFields = getFilter(config, 'fields')
+  const configMetadata = getFilter(config, 'metadata')
+
+  function datadogInterceptor (options, nextCall) {
+    const path = options.method_definition.path
+    const methodType = getMethodType(options.method_definition)
+    const scope = tracer.scopeManager().active()
+    const span = tracer.startSpan('grpc.request', {
+      childOf: scope && scope.span(),
+      tags: {
+        [Tags.SPAN_KIND]: 'client',
+        'grpc.method.name': path,
+        'grpc.method.type': methodType,
+        'resource.name': path,
+        'service.name': config.service || `${tracer._service}-grpc-client`
+      }
+    })
+
+    return new grpc.InterceptingCall(nextCall(options), {
+      sendMessage: (message, next) => {
+        if (configFields) {
+          const values = configFields(message)
+
+          for (const key in values) {
+            span.setTag(`grpc.request.message.fields.${key}`, values[key])
```

Comment:
What happens if there are multiple messages in the request? It looks like this should be attached to the message span but it doesn't exist in this implementation.

@nunofgs Is this something you need? If not I would say to just leave it out for this initial implementation. Otherwise a new message span will probably need to be introduced to hold these tags.

The same applies to the server as well.

Reply:
Not sure I know what you mean. A span is being created upon each grpc message, both on the client and server. Maybe you mean something else?

Reply:
My understanding from other languages is that there is a single request (or stream) for multiple messages. A span is created on the stream and may contain multiple messages, which may or may not have spans, but since there is a single span created here I assumed it was for the stream and not the message, especially since it handles propagation. Is this not correct?

Reply:
🤔 from what I can see, client interceptors are called on every message. Even if there are multiple messages per request, a client interceptor call is scoped to a single method which is where the span is being created, right?

Reply:
I think the safest would be to have a test for streams with multiple messages. This will make it clear what the behavior is and will ensure that the implementation hooks in the right place when there are multiple messages.

In other languages we use different spans for the request and the messages, so this is important to ensure interoperability.